src/main/groovy/org/pih/warehouse/importer/CycleCountItemsExcelImporter.groovy

```diff
@@ -42,17 +46,43 @@ class CycleCountItemsExcelImporter extends AbstractExcelImporter implements Data
         super()
         read(fileName)
         excelImportService = Holders.grailsApplication.mainContext.getBean("excelImportService")
-
     }
 
     @Override
     List<Map> getData() {
-        excelImportService.columns(
+        List<Map> data = excelImportService.columns(
                 workbook,
                 COLUMN_MAP,
                 CELL_REPORTER,
                 PROPERTY_MAP
-        )
+        ) as List<Map>
+
+        // 1. Collect all uniques bin location names and product codes
+        List<String> locationNames = data*.binLocation.unique()
+        List<String> productCodes  = data*.productCode.unique()
+
+        // 2. Fetch all of the necessary data in single database call
+        Map<String, Location> locationMap = Location.findAllByNameInList(locationNames).collectEntries {
+            [it.name, it]
+        }
+        Map<String, Product> productMap = Product.findAllByProductCodeInList(productCodes).collectEntries {
+            [it.productCode, it]
+        }
+
+        // 3. Replace data from importer with the fetched data
+        data.each { row ->
+            row.binLocation = [
+                    id: locationMap[row.binLocation]?.id,
+                    name: locationMap[row.binLocation]?.name
+            ]
+            row.product = [
+                    id: productMap[row.productCode]?.id,
+                    name: productMap[row.productCode]?.name,
+                    productCode: productMap[row.productCode]?.productCode
+            ]
+        }
+
+        return data
```

Comment:
I added fetching locations and products in one database call to avoid performance issues. At this moment, I tried importing 200 rows. Getting the response took < 500ms, rendering lines on the frontend side took 1s

Reply:
nice good idea :+1: :+1: :+1: